File: app/src/main/java/com/coinninja/coinkeeper/util/CurrencyPreference.java
package com.coinninja.coinkeeper.util;

import android.content.Intent;

import com.coinninja.coinkeeper.factory.CurrencyFactory;
import com.coinninja.coinkeeper.util.android.LocalBroadCastUtil;
import com.coinninja.coinkeeper.util.android.PreferencesUtil;
import com.coinninja.coinkeeper.util.currency.BTCCurrency;
import com.coinninja.coinkeeper.util.currency.Currency;
import com.coinninja.coinkeeper.util.currency.USDCurrency;

import javax.inject.Inject;

public class CurrencyPreference {
    static final String PREFERENCE_PRIMARY_CURRENCY = "primaryCurrency";
    static final String PREFERENCE_SECONDARY_CURRENCY = "secondaryCurrency";
    private final PreferencesUtil preferencesUtil;
    private final CurrencyFactory currencyFactory;
    private final LocalBroadCastUtil localBroadCastUtil;

    @Inject
    public CurrencyPreference(PreferencesUtil preferencesUtil, CurrencyFactory currencyFactory,
                              LocalBroadCastUtil localBroadCastUtil) {
        this.preferencesUtil = preferencesUtil;
        this.currencyFactory = currencyFactory;
        this.localBroadCastUtil = localBroadCastUtil;
    }

    public void reset() {
        currencyFactory.reset();
    }

    public void setCurrencies(Currency primary, Currency secondary) {
        preferencesUtil.savePreference(PREFERENCE_PRIMARY_CURRENCY, primary.getSymbol());
        preferencesUtil.savePreference(PREFERENCE_SECONDARY_CURRENCY, secondary.getSymbol());
        Intent intent = new Intent(Intents.ACTION_CURRENCY_PREFERENCE_CHANGED);
        intent.putExtra(Intents.EXTRA_PREFERENCE, getCurrenciesPreference());
        localBroadCastUtil.sendBroadcast(intent);
    }

    public DefaultCurrencies getCurrenciesPreference() {
        String primarySymbol = preferencesUtil.getString(PREFERENCE_PRIMARY_CURRENCY, BTCCurrency.SYMBOL);
        String secondarySymbol = preferencesUtil.getString(PREFERENCE_SECONDARY_CURRENCY, USDCurrency.SYMBOL);
        Currency primaryCurrency = currencyFactory.fromSymbol(primarySymbol);
        Currency secondaryCurrency = currencyFactory.fromSymbol(secondarySymbol);
        return new DefaultCurrencies(primaryCurrency, secondaryCurrency);
    }

    public DefaultCurrencies toggleDefault() {
        DefaultCurrencies defaultCurrencies = getCurrenciesPreference();
        setCurrencies(defaultCurrencies.getSecondaryCurrency(), defaultCurrencies.getPrimaryCurrency());
        return new DefaultCurrencies(defaultCurrencies.getSecondaryCurrency(), defaultCurrencies.getPrimaryCurrency());
    }

    public Currency getFiat() {
        return getCurrenciesPreference().getFiat();
    }
}

File: app/src/main/java/com/coinninja/coinkeeper/util/CurrencyPreference.kt
package com.coinninja.coinkeeper.util

import android.content.Intent
import app.dropbit.annotations.Mockable
import app.dropbit.commons.currency.BTCCurrency
import app.dropbit.commons.currency.Currency
import app.dropbit.commons.currency.USDCurrency
import com.coinninja.coinkeeper.factory.CurrencyFactory
import com.coinninja.coinkeeper.util.android.LocalBroadCastUtil
import com.coinninja.coinkeeper.util.android.PreferencesUtil
import javax.inject.Inject

@Mockable
class CurrencyPreference @Inject constructor(internal val preferencesUtil: PreferencesUtil,
                                             internal val currencyFactory: CurrencyFactory,
                                             internal val localBroadCastUtil: LocalBroadCastUtil) {


    val currenciesPreference: DefaultCurrencies
        get() {
            val primarySymbol = preferencesUtil.getString(PREFERENCE_PRIMARY_CURRENCY, USDCurrency.SYMBOL)
            val secondarySymbol = preferencesUtil.getString(PREFERENCE_SECONDARY_CURRENCY, BTCCurrency.SYMBOL)
            val primaryCurrency = currencyFactory.fromSymbol(primarySymbol)
            val secondaryCurrency = currencyFactory.fromSymbol(secondarySymbol)
            return DefaultCurrencies(primaryCurrency, secondaryCurrency)
        }

    val fiat: Currency get() = currenciesPreference.fiat

    fun reset() = currencyFactory.reset()

    fun toggleDefault(): DefaultCurrencies {
        val defaultCurrencies = currenciesPreference
        setCurrencies(defaultCurrencies.secondaryCurrency, defaultCurrencies.primaryCurrency)
        return DefaultCurrencies(defaultCurrencies.secondaryCurrency, defaultCurrencies.primaryCurrency)
    }

    internal fun setCurrencies(primary: Currency, secondary: Currency) {
        preferencesUtil.savePreference(PREFERENCE_PRIMARY_CURRENCY, primary.symbol)
        preferencesUtil.savePreference(PREFERENCE_SECONDARY_CURRENCY, secondary.symbol)
        val intent = Intent(DropbitIntents.ACTION_CURRENCY_PREFERENCE_CHANGED)
        intent.putExtra(DropbitIntents.EXTRA_PREFERENCE, currenciesPreference)
        localBroadCastUtil.sendBroadcast(intent)
    }

    companion object {
        internal const val PREFERENCE_PRIMARY_CURRENCY = "primaryCurrency"
        internal const val PREFERENCE_SECONDARY_CURRENCY = "secondaryCurrency"
    }
}
